plugins/inputs/opensearch_query/opensearch_query.go

```diff
@@ -0,0 +1,326 @@
+	// retrieve field mapping and build queries only once
+	agg.mapMetricFields, err = o.getMetricFields(ctx, agg)
+	if err != nil {
+		return fmt.Errorf("not possible to retrieve fields: %v", err.Error())
+	}
+
+	for _, metricField := range agg.MetricFields {
+		if _, ok := agg.mapMetricFields[metricField]; !ok {
+			return fmt.Errorf("metric field '%s' not found on index '%s'", metricField, agg.Index)
+		}
+	}
+
+	err = agg.buildAggregationQuery()
+	if err != nil {
+		return fmt.Errorf("error building aggregation: %s", err)
+	}
+
+	o.Aggregations[i] = agg
+	return nil
+}
+
+func (o *OpensearchQuery) connectToOpensearch() error {
+	var client *opensearch.Client
+	var transport *http.Transport
+
+	if o.InsecureSkipVerify {
+		transport = &http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+		}
+	}
+
+	clientConfig := opensearch.Config{
+		Addresses: o.URLs,
+		Username:  o.Username,
+		Password:  o.Password,
+	}
+
+	if transport != nil {
+		clientConfig.Transport = transport
+	}
+
+	client, err := opensearch.NewClient(clientConfig)
+	o.osClient = client
+
+	return err
+}
+
+// Gather writes the results of the queries from OpenSearch to the Accumulator.
+func (o *OpensearchQuery) Gather(acc telegraf.Accumulator) error {
+	var wg sync.WaitGroup
+
+	for _, agg := range o.Aggregations {
+		wg.Add(1)
+		go func(agg osAggregation) {
+			defer wg.Done()
+			err := o.osAggregationQuery(acc, agg)
+			if err != nil {
+				acc.AddError(fmt.Errorf("opensearch query aggregation %s: %s ", agg.MeasurementName, err.Error()))
+			}
+		}(agg)
+	}
+
+	wg.Wait()
+	return nil
+}
+
+func (o *OpensearchQuery) osAggregationQuery(acc telegraf.Accumulator, aggregation osAggregation) error {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(o.Timeout))
+	defer cancel()
+
+	searchResult, err := o.runAggregationQuery(ctx, aggregation)
+	if err != nil {
+		return err
+	}
+
+	return searchResult.GetMetrics(acc, aggregation.MeasurementName)
+}
+
+func init() {
+	inputs.Add("opensearch_query", func() telegraf.Input {
+		return &OpensearchQuery{
+			Timeout:             config.Duration(time.Second * 5),
+			HealthCheckInterval: config.Duration(time.Second * 10),
+		}
+	})
+}
+
+func (o *OpensearchQuery) runAggregationQuery(ctx context.Context, aggregation osAggregation) (*AggregationResponse, error) {
+	now := time.Now().UTC()
+	from := now.Add(time.Duration(-aggregation.QueryPeriod))
+	filterQuery := aggregation.FilterQuery
+	if filterQuery == "" {
+		filterQuery = "*"
+	}
+
+	aq := &Query{
+		Size:         0,
+		Aggregations: aggregation.aggregation,
+		Query:        nil,
+	}
+
+	boolQuery := &BoolQuery{
+		FilterQueryString: filterQuery,
+		TimestampField:    aggregation.DateField,
+		TimeRangeFrom:     from,
+		TimeRangeTo:       now,
+		DateFieldFormat:   aggregation.DateFieldFormat,
+	}
+
+	aq.Query = boolQuery
+	req, err := json.Marshal(aq)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal request: %s", err)
```

Comment:
Please wrap the error if we ever want to reason about those upstream...
```suggestion
		return nil, fmt.Errorf("failed to marshal request: %w", err)
```